README
NAME
    Dist::Zilla::Plugin::Bootstrap::lib - A minimal boot-strapping for
    Dist::Zilla Plug-ins.

VERSION
    version 1.001001

SYNOPSIS
        [Bootstrap::lib]
        try_built   = 1  ; try using an existing built distribution named Dist-Name-*
        fallback    = 0  ; if try_built can't find a built distribution, or there's more than one, don't bootstrap
                         ; using lib/ instead

DESCRIPTION
    This module exists for loading either "/lib" or
    "/Dist-Name-$VERSION/lib" into your @INC early on.

    This is mostly useful for writing "Dist::Zilla" plug-ins, so that you
    may build and release a plug-in using itself.

USE CASES
  Simple single-phase self-dependency
    This module really is only useful in the case where you need to use
    something like

        dzil -Ilib

    For *every* call to "Dist::Zilla", and this is mainly a convenience.

    For that

        [Bootstrap::lib]

    on its own will do the right thing.

  Installed-Only self-dependency
    The other useful case is when you would normally do

        dzil build                 # pass 1 that generates Foo-1.234 with a pre-installed Foo-1.233
        dzil -IFoo-1.234/lib build # pass 2 that generates Foo-1.234 with Foo-1.234

    For that

        [Bootstap::lib]
        try_built   = 1
        fallback    = 0

    Will do what you want.

        dzil build   # pass1 -> creates Foo-1.234 without bootstrapping
        dzil build   # pass2 -> creates Foo-1.234 boot-strapped from the previous build

  2-step self-dependency
    There's a 3rd useful case which is a hybrid of the 2, where you /can/
    build from your own sources without needing a pre-installed version,
    just you don't want that for release code ( e.g.: $VERSION being "undef"
    in code that is run during release is "bad" )

        [Bootstrap::lib]
        try_built = 1
        fallback  = 1

    Then

        dzil build  # pass1 -> creates Foo-1.234 from bootstrapped $root/lib
        dzil build  # pass2 -> creates Foo-1.234 from bootstrapped $root/Foo-1.234

PRECAUTIONS
  DO NOT
    DO NOT use this library from inside a bundle. It will not likely work as
    expected, and you DO NOT want to bootstrap everything in all cases.

  NO VERSION
    On its own,

        [Bootstrap::lib]

    At present, using this module in conjunction with a module with no
    explicitly defined version in the source will result in the *executed*
    instance of that plug-in *also* having NO VERSION.

    If this is a problem for you, then its suggested you try either
    variation of using

        [Bootstrap::lib]
        try_built = 1
        ; no_fallback = 1   #

  SUCKS AT GUESSING
    The core mechanism behind "try_built" relies on looking in your project
    directory for a previous build directory of some kind.

    And there's no way for it to presently pick a "best" version when there
    are more than one, or magically provide a better solution if there are
    "zero" versions readily available.

    This is mostly because there is no way to determine the "current"
    version we are building for, because the point in the execution cycle is
    so early, no version plugins are likely to be even instantiated yet, and
    some version plugins are dependent on incredibly complex precursors (
    like git ), so by even trying to garner the version we're currently
    building, we could be prematurely cutting off a vast majority of modules
    from even being able to bootstrap.

    Even as it is, us using "zilla->name" means that if your dist relies on
    some process to divine its name, the module that does this must:

    *   be loaded and declared prior to "Bootstrap::lib" in the "dist.ini"

    *   not itself be the module you are presently developing/bootstrapping

    The only way of working around that I can envision is adding parameters
    to "Bootstrap::lib" to specify the dist name and version name... but if
    you're going to do that, you may as well stop using external plugins to
    discover that, and hard-code those values in "dist.ini" to start with.

  STILL NOT REALLY A PLUGIN
    Starting at version 0.04000000 ( read: 0.04_0000_00 aka 0.04 + 0 x 4 + 0
    x 2 ) this module is a fully fledged class, different only from standard
    Dist::Zilla Plugins in that it doesn't partake in normal phase order,
    and only executes during a special custom "::Bootstrap" phase, which is
    more or less a different name and implementation of "BUILD", in that
    "bootstrap" is invoked after "plugin_from_config" is called ( where
    "new" is called ), which occurs somewhere in the middle of
    "register_component"

    This module also appears on the plugin stash, and responds naturally to
    "metaconfig" requests.

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric <kentnl@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Dist-Zilla-Plugin-Bootstrap-lib

1.001002 2017-03-18T04:56:00Z 0a1cb3d
 - Stop using Test::KENTNL/dztest

 [Dependencies::Stats]
 - Dependencies changed since 1.001001, see misc/*.deps* for details
 - configure: ↓1
 - develop: ↑1 (suggests: ↑2)
 - runtime: ↓1 -1
 - test: ↓1 -1

1.001001 2017-03-01T21:09:23Z d39d30d
 [Bugfix]
 - Update tooling to avoid .-missing-in-INC bugs in tests
 - Closes https://github.com/kentnl/Dist-Zilla-Plugin-Bootstrap-lib/issues/3

 [Dependencies::Stats]
 - Dependencies changed since 1.001000, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +13 ↑3 -3 (recommends: +1, suggests: ↑2)
 - runtime: (recommends: +1)
 - test: +1 ↓1 (recommends: ↑1 ↓1)

1.001000 2014-08-31T19:36:06Z
 [00 Minor]
 - Test refactor.

 [Dependencies::Stats]
 - Dependencies changed since 1.000005, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)
 - test: +1 -5

 [Tests]
 - Reimplement tests using dztest.

1.000005 2014-08-15T21:16:17Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - whitespace reduction.

 [Dependencies::Stats]
 - Dependencies changed since 1.000004, see misc/*.deps* for details
 - develop: +1 (suggests: ↑1)

 [Misc]
 - Improve source side POD to reduce build side whitespace.

1.000004 2014-08-03T12:29:48Z
 [Dependencies::Stats]
 - Dependencies changed since 1.000003, see misc/*.deps* for details
 - configure: ↓1
 - develop: (suggests: ↑1)
 - test: +1 -3 (recommends: ↑1 -1)

 [Misc]
 - use -SingleEncoding
 - Get newer Test::ReportPrereqs

1.000003 2014-07-15T12:39:08Z
 [00 Trivial]
 - use newer Role::Bootstrap ( closes #2 )

 [Dependencies::Stats]
 - Dependencies changed since 1.000002, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - runtime: ↑1

1.000002 2014-07-15T07:33:24Z
 [00 Trivial]
 - Misc packaging cleanups

 [Dependencies::Stats]
 - Dependencies changed since 1.000001, see misc/*.deps* for details
 - configure: (recommends: ↑1)
 - develop: +3 -3 (suggests: +2 -1)
 - test: (recommends: ↑1)

1.000001 2014-04-05T11:36:12Z
 [00 Test Fixes]
 - Drop dependency on Path::FindDev

 [Dependencies::Aggregate]
 - 5.008 : 207 → 204
 - 5.010 : 186 → 183
 - 5.019 : 167 → 164

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - test: +1 -1

1.000000 2014-04-05T06:23:20Z
 [00 Major]
 - This release contains mostly packaging changes that will be invisible to CPAN users.
 - However, it may require attention for downstream non-CPAN tooling.

 [01 Version Scheme]
 - Version Scheme is shortened from a mantissa of 8 to a mantissa of 6
 - This is the primary reason for the 1.xxx release.

 [02 Tooling is MakeMaker based]
 - Tooling is now switched to ExtUtils::MakeMaker
 - instead of coercing users into installing the latest EUMM/TestMore/MB when not even necessary, that is now only the
   recommended course of action.

 [Dependencies::Aggregate]
 - 5.008: 204 → 207
 - 5.010: 184 → 186
 - 5.019: 168 → 167

 [Dependencies::Stats]
 - Dependencies changed since 0.04000002, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +7 ↑1 -9
 - runtime: +1 ↑1
 - test: +4 ↓1 (recommends: +4)

0.04000002 2013-11-18T07:22:31Z
 [00 Minor/Bugfix]
 - Regenerate files to try avoid a few failures
 - Use Compile::PerFile
 - Use latest Module::Build magic
 - Rewrite tests to avoid stacked operators for 5.8 compat.

 [Dependencies::Stats]
 - Dependencies changed since 0.04000001, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +43 ↑1 -1 (recommends: -1, suggests: +1 -1)
 - test: ↑1 -3

 [Documentation]
 - README.mkdn points to metacpan

0.04000001 2013-09-24T16:22:46Z
 [Bugfix]
 - Attempt to solve the problem of CPAN Smokers not having dzil in path by using Test::DZil instead.

 [Dependencies::Stats]
 - Dependencies changed since 0.04000000, see misc/*.deps* for details
 - test: +3 -1

 [Tests]
 - additional test for new ::Role::Bootstrap behaviour
 - update compile test

0.04000000 2013-09-05T08:17:38Z
 [!Major and API breaking Changes]
 - no_fallback deprecated in favour of fallback = 0/1
 - reimplemented in terms of a proper Moose class and ::Role::Bootstrap

 [!Minor]
 - add metapod and related info

 [Dependencies::Stats]
 - Dependencies changed since 0.03000200, see misc/*.deps* for details
 - runtime: +2 -5
 - test: +2

0.03000200 2013-09-03T03:36:10Z
 [!Minor]
 - Add tests
 - Minor refactor

 [Dependencies::Stats]
 - Dependencies changed since 0.03000101, see misc/*.deps* for details
 - test: +3

0.03000101 2013-08-31T01:52:46Z
 [Bugfix]
 - accidental typo with boostrap_dir instead of bootstrap_dir

0.03000100 2013-08-31T01:43:36Z
 [!Minor]
 - Now shows config in metaconfig with a bit of Moo wizzardry.

 [Dependencies::Stats]
 - Dependencies changed since 0.03000000, see misc/*.deps* for details
 - runtime: +2

0.03000000 2013-08-30T23:46:07Z
 [!Major]
 - Less noisy in general usage.
 - Now warns if paths that would be boostraped contain modules that are already loaded in %INC ( closes gh #1 )
 - Tests regenerated with newer tools

 [Dependencies::Stats]
 - Dependencies changed since 0.02000100, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +2 (suggests: ↑1)
 - runtime: +1
 - test: +2 -2

0.02000100 2013-04-06T09:53:45Z
 [Bugs]
 - 2 fallback conditions were accidentally using the same logic, so only one would ever work.
 - satisfactory build dirs accidentally lacked a 'lib' suffix, so resulted in a "fallthrough" to system.

0.02000000 2013-04-06T09:38:52Z
 [!Noteworthy Changes]
 - @INC mutation now done slightly later at C<register_component>, which lets us get payload parameters out, while
   still executing the real work long before anything else. Though execution order still matters profusely.
 - 3 new parameters,
 - `try_built` tries to use a previous build instead of a "undef version" based bootstrap

Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.008.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4203"
  },
  "configure_requires" => {
    "Module::Build" => "0.4203"
  },
  "dist_abstract" => "A minimal boot-strapping for Dist::Zilla Plug-ins.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Dist-Zilla-Plugin-Bootstrap-lib",
  "dist_version" => "0.04000003",
  "license" => "perl",
  "module_name" => "Dist::Zilla::Plugin::Bootstrap::lib",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Dist::Zilla::Role::Bootstrap" => 0,
    "Moose" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Cwd" => 0,
    "File::Copy::Recursive" => 0,
    "Path::FindDev" => 0,
    "Path::Tiny" => 0,
    "Test::DZil" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "1.001002"
  }
);


my %fallback_build_requires = (
  "Cwd" => 0,
  "File::Copy::Recursive" => 0,
  "Module::Build" => "0.4203",
  "Path::FindDev" => 0,
  "Path::Tiny" => 0,
  "Test::DZil" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "1.001002"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# FILENAME: bundle_to_ini.pl
# CREATED: 02/06/14 01:48:56 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Write an INI file from a bundle

use 5.008;    #utf8
use strict;
use warnings;
use utf8;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    pre
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  open my $rcfile, '>', './perlcritic.rc' or croak 'Cant open perlcritic.rc';
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  open my $depsfile, '>', './perlcritic.deps' or croak 'Cant open perlcritic.deps';
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A minimal boot-strapping for Dist::Zilla Plug-ins.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dist-Zilla-Plugin-Bootstrap-lib",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "Dist::Zilla::Plugin::Bootstrap::lib",
  "PREREQ_PM" => {
    "Dist::Zilla::Role::Bootstrap" => "1.000000",
    "Moose" => 0,
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::DZil" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.001002",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Dist::Zilla::Role::Bootstrap" => "1.000000",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Moose" => 0,
  "Test::DZil" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
